interface/ceed-qfunctioncontext.c

```diff
@@ -0,0 +1,306 @@
+  @param[out] ctx   Address of the variable where the newly created
+                      CeedQFunctionContext will be stored
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedQFunctionContextCreate(Ceed ceed, CeedQFunctionContext *ctx) {
+  int ierr;
+
+  if (!ceed->QFunctionContextCreate) {
+    Ceed delegate;
+    ierr = CeedGetObjectDelegate(ceed, &delegate, "Context"); CeedChk(ierr);
+
+    if (!delegate)
+      // LCOV_EXCL_START
+      return CeedError(ceed, 1, "Backend does not support ContextCreate");
+    // LCOV_EXCL_STOP
+
+    ierr = CeedQFunctionContextCreate(delegate, ctx); CeedChk(ierr);
+    return 0;
+  }
+
+  ierr = CeedCalloc(1, ctx); CeedChk(ierr);
+  (*ctx)->ceed = ceed;
+  ceed->refcount++;
+  (*ctx)->refcount = 1;
+  ierr = ceed->QFunctionContextCreate(*ctx); CeedChk(ierr);
+  return 0;
+}
+
+/**
+  @brief Set the data used by a CeedQFunctionContext, freeing any previously allocated
+           data if applicable. The backend may copy values to a different
+           memtype, such as during @ref CeedQFunctionApply().
+           See also @ref CeedQFunctionContextTakeData().
+
+  @param ctx   CeedQFunctionContext
+  @param mtype Memory type of the data being passed
+  @param cmode Copy mode for the data
+  @param data  Data to be used
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedQFunctionContextSetData(CeedQFunctionContext ctx, CeedMemType mtype,
+                                CeedCopyMode cmode,
+                                size_t size, void *data) {
+  int ierr;
+
+  if (!ctx->SetData)
+    // LCOV_EXCL_START
+    return CeedError(ctx->ceed, 1, "Backend does not support ContextSetData");
+  // LCOV_EXCL_STOP
+
+  if (ctx->state % 2 == 1)
+    // LCOV_EXCL_START
+    return CeedError(ctx->ceed, 1,
+                     "Cannot grant CeedQFunctionContext data access, the "
+                     "access lock is already in use");
+  // LCOV_EXCL_STOP
+
+  ctx->ctxsize = size;
+  ierr = ctx->SetData(ctx, mtype, cmode, data); CeedChk(ierr);
+  ctx->state += 2;
+
+  return 0;
+}
+
+/**
+  @brief Get read/write access to a CeedQFunctionContext via the specified memory type.
+           Restore access with @ref CeedQFunctionContextRestoreData().
+
+  @param ctx        CeedQFunctionContext to access
+  @param mtype      Memory type on which to access the data. If the backend
+                    uses a different memory type, this will perform a copy.
+  @param[out] data  Data on memory type mtype
+
+  @note The CeedQFunctionContextGetData() and @ref CeedQFunctionContextRestoreData() functions
+    provide access to array pointers in the desired memory space. Pairing
+    get/restore allows the Context to track access.
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref User
+**/
+int CeedQFunctionContextGetData(CeedQFunctionContext ctx, CeedMemType mtype,
+                                void **data) {
```

Comment:
Needs Fortran usage notes. Also, consider using `void *` (like we do in `CeedMallocArray`) instead of `void **`.

Reply:
I added better usage notes to `CeedQFunctionGetContext` and `CeedQFunctionGetInnerContext`

Re `void *` vs `void **`, we use `void **` quite a bit across the library. I'm not sure I understand why we do so

Reply:
So `void **` has the right arity: we pass the address of a pointer to (unknown to us) data. However, `void **` always needs a cast because `Thing **` does not implicitly convert to `void *` so it *always* needs an explicit cast. So other libraries, including MPI (cf. `MPI_Comm_get_attr`) and PETSc, just use `void *` and document that one should pass the address of the pointer. C and C++ implicitly converts to `void *`, so no cast is necessary. (C implicitly converts _from_ `void *` while C++ does not.)

Reply:
That's what I thought. So should we just pick one and stick with it for the whole library? It would be tedious but easy to change over.

Reply:
I refer to avoid casts except in unusual circumstances so I'd be in favor of using `void *` for this purpose despite it having the "wrong" arity.

Reply:
Ok, should be good now.